rfc/src/rfcs/0009-function-contracts.md

```diff
@@ -0,0 +1,845 @@
+    unsafe { IS_ENTERED = true; }
+    let result = check_pop(self);
+    unsafe { IS_ENTERED = false; }
+    result
+  };
+}
+```
+
+Note that this is insufficient to verify all types of recursive functions, as
+the contract specification language has no support for inductive lemmas (for
+instance in [ACSL](https://frama-c.com/download/acsl.pdf) section 2.6.3
+"inductive predicates"). Inductive lemmas are usually needed for recursive
+data structures.
+
+### Changes to Other Components
+
+Contract enforcement and replacement (`kani::proof_for_contract(f)`,
+`kani::stub_verified(f)`) both dispatch to the **stubbing logic**, stubbing `f`
+with the generated check and replace function respectively. If `f` has no
+contract, Kani throws an error.
+
+For **write sets** Kani relies on CBMC. `modifies` clauses (whether derived from
+types or from explicit clauses) are emitted from the compiler as GOTO contracts
+in the artifact. Then the driver invokes `goto-instrument` with the name of the
+GOTO-level function names to enforce or replace the memory contracts. The
+compiler communicates the names of the function via harness metadata.
+
+
+Code used in contracts is required to be **side effect** free which means it
+must not perform I/O, mutate memory (`&mut` vars and such) or (de)allocate heap
+memory. This is enforced in two layers. First with an MIR traversal over all
+code reachable from a contract expression. An error is thrown if known
+side-effecting actions are performed such as `ptr::write`, `malloc`, `free` or
+functions which we cannot check, such as e.g. `extern "C"`, with the exception
+of known side effect free functions in e.g. the standard library.
+
+<!-- 
+This is the technical portion of the RFC. Please provide high level details of the implementation you have in mind:
+
+- What are the main components that will be modified? (E.g.: changes to `kani-compiler`, `kani-driver`, metadata,
+  installation...)
+- How will they be modified? Any changes to how these components communicate?
+- Will this require any new dependency?
+- What corner cases do you anticipate? 
+-->
+
+## Rationale and alternatives
+
+<!-- For Developers -->
+<!-- `old` discussion here -->
+
+<!-- 
+- What are the pros and cons of this design?
+- What is the impact of not doing this?
+- What other designs have you considered? Why didn't you choose them? 
+-->
+
+
+### Kani-side implementation vs CBMC 
+
+Instead of generating check and replace functions in Kani, we could use the contract instrumentation provided by CBMC.
+
+We tried this earlier but came up short, because it is difficult to implement,
+while supporting arbitrary Rust syntax. We exported the conditions into
+functions so that Rust would do the parsing/type checking/lowering for us and
+then call the lowered function in the CBMC contract. 
+
+The trouble is that CBMC's `old` is only supported directly in the contract, not
+in functions called from the contract. This means we either need to inline the
```

Comment:
this is to avoid context capture problems that are discussed above

Reply:
Absolutely. I don't see this as a judgement on CBMC, after all we have the exact same limitation. But it makes it more difficult to use for Kani.